core/src/main/java/org/geysermc/geyser/translator/text/MessageTranslator.java

```diff
@@ -426,17 +431,91 @@ public static String normalizeSpace(String string) {
-            return "";
+        Component parsed = componentFromNbtTag(description);
+        return convertMessage(session, parsed);
+    }
+
+    public static Component componentFromNbtTag(Object nbtTag) {
+        return componentFromNbtTag(nbtTag, Style.empty());
+    }
+
+    private static Component componentFromNbtTag(Object nbtTag, Style style) {
+        if (nbtTag instanceof String literal) {
+            return Component.text(literal).style(style);
+        } else if (nbtTag instanceof List<?> list) {
+            return Component.join(JoinConfiguration.noSeparators(), componentsFromNbtList(list, style));
+        } else if (nbtTag instanceof NbtMap map) {
+            Component component = null;
+            String text = map.getString("text", null);
+            if (text != null) {
+                component = Component.text(text);
+            } else {
+                String translateKey = map.getString("translate", null);
+                if (translateKey != null) {
+                    String fallback = map.getString("fallback", "");
+                    List<Component> args = new ArrayList<>();
+
+                    Object with = map.get("with");
+                    if (with instanceof List<?> list) {
+                        args = componentsFromNbtList(list, style);
+                    } else if (with != null) {
+                        args.add(componentFromNbtTag(with, style));
+                    }
+                    component = Component.translatable(translateKey, fallback, args);
+                }
+            }
+
+            if (component != null) {
+                Style newStyle = getStyleFromNbtMap(map, style);
+                component = component.style(newStyle);
+
+                Object extra = map.get("extra");
+                if (extra != null) {
+                    component = component.append(componentFromNbtTag(extra, newStyle));
+                }
+
+                return component;
+            }
         }
-        return translate;
+
+        throw new IllegalArgumentException("Expected tag to be a literal string, a list of components, or a component object with a text/translate key");
+    }
+
+    private static List<Component> componentsFromNbtList(List<?> list, Style style) {
+        List<Component> components = new ArrayList<>();
+        for (Object entry : list) {
+            components.add(componentFromNbtTag(entry, style));
+        }
+        return components;
+    }
+
+    public static Style getStyleFromNbtMap(NbtMap map) {
+        Style.Builder style = Style.style();
+
+        String colorString = map.getString("color", null);
+        if (colorString != null) {
+            if (colorString.startsWith(TextColor.HEX_PREFIX)) {
+                style.color(TextColor.fromHexString(colorString));
+            } else {
+                style.color(NamedTextColor.NAMES.value(colorString));
+            }
+        }
```

Comment:
could probably use `listenForString` if going that route with the others

Reply:
I'm not too concerned about this since it's one map call and a lambda isn't strictly necessary to achieve something that works fine as-is. `listenForBoolean` is different because it's a primitive object.

Reply:
aahhh right, the latter avoids boxing